core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -62,17 +46,22 @@ trait MetadataCache {
 
   def getAllTopics(): collection.Set[String]
 
-  def getAllPartitions(): collection.Set[TopicPartition]
+  def getTopicPartitions(topicName: String): collection.Set[TopicPartition]
 
-  def getNonExistingTopics(topics: collection.Set[String]): collection.Set[String]
+  def hasAliveBroker(brokerId: Int): Boolean
 
-  def getAliveBroker(brokerId: Int): Option[MetadataBroker]
+  def getAliveBrokers(): Iterable[BrokerMetadata]
 
-  def getAliveBrokers: collection.Seq[MetadataBroker]
+  def getAliveBrokerNode(brokerId: Int, listenerName: String): Option[Node]
+
+  def getAliveBrokerNodes(listenerName: String): Iterable[Node]
 
   def getPartitionInfo(topic: String, partitionId: Int): Option[UpdateMetadataRequestData.UpdateMetadataPartitionState]
 
-  def numPartitions(topic: String): Option[Int]
+  /**
+   * Return the number of partitions in the given topic, or 0 if the given topic does not exist.
+   */
+  def numPartitions(topic: String): Int
```

Comment:
It just seems awkward because typically when you ask "how many partitions?" the answer is not Some or None, but a number.

Reply:
I agree that the Option rather than a special value is better here. It's not a big deal either way for the current code since there aren't many usages, but in the future a new caller will have to know to check for `0`. Seems like it could be easy to introduce bugs.

A longer term solution (beyond the scope of this PR) could be to have individual container objects for the topic-level metadata. This way we could easily encapsulate the "does this topic even exist" with an Option on the metadata cache. E.g., `metadataCache.topic("foo").map(_.numPartitions)`

Reply:
OK, I will just convert it back to an `Option[Int]`, then.